modules/s3-articfact/main.tf

```
@@ -0,0 +1,380 @@
+# - Primary IAM Policy Document
+# - Primary IAM Policy
+# - Replica IAM Policy Document
+# - Replica IAM Role
+# - Replica IAM Role Policy Attachment
+#
+# -------------------------------------------------------------------------------------
+
+
+# -------------------------------------------
+# SET TERRAFORM REQUIREMENTS TO RUN MODULE
+# -------------------------------------------
+
+terraform {
+  required_version = ">= 1.5.5"
+
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = ">= 5.0"
+    }
+  }
+}
+
+# -------------------------------------------
+# SET PROVIDER ALIAS FOR REPLICATION
+# -------------------------------------------
+
+provider "aws" {
+  region = var.priamry_region
+}
+
+provider "aws" {
+  alias  = "replica"
+  region = var.replica_region
+}
+
+
+# ------------------------------------------------------------
+
+# THE FOLLOWING SECTION IS USED TO CREATE THE PRIMARY S3 BUCKET
+
+# AND ASSOCIATED RESOURCES (LIFECYCLE MANAGEMENT, VERSIONING).
+
+# ------------------------------------------------------------
+
+
+# -------------------------------------------
+# CREATE PRIMARY S3 BUCKET
+# -------------------------------------------
+
+resource "aws_s3_bucket" "primary" {
+  bucket = var.primary_bucket_name
+}
+
+# -------------------------------------------
+# CREATE PRIMARY S3 BUCKET POLICY
+# -------------------------------------------
+
+resource "aws_s3_bucket_acl" "primary" {
+  bucket = aws_s3_bucket.primary.id
+  acl    = var.pimary_bucket_acl
+}
+
+# -------------------------------------------
+# ENABLE PRIMARY OBJECT VERSIONING
+# -------------------------------------------
+
+resource "aws_s3_bucket_versioning" "primary" {
+  bucket = aws_s3_bucket.primary.id
+  versioning_configuration {
+    status = var.bucket_versioning_configuration_status
+  }
+}
+
+# -------------------------------------------
+# CONFIGURE PARTIAL LIFECYCLE MANAGEMENT
+# -------------------------------------------
+
+resource "aws_s3_bucket_lifecycle_configuration" "pirmary" {
+  count = !var.enable_storage_lifecycles ? 1 : 0
+
+  bucket = aws_s3_bucket.primary.id
+
+  rule {
+    id = var.lifecycle_versioning_id
+
+    status = var.bucket_versioning_configuration_status
+
+    noncurrent_version_expiration {
+      noncurrent_days           = 30
+      newer_noncurrent_versions = 1
+    }
+  }
+}
+
+# -------------------------------------------
+# CONFIGURE FULL LIFECYCLE MANAGEMENT
+# -------------------------------------------
+
+resource "aws_s3_bucket_lifecycle_configuration" "primary" {
+  count = var.full_lifecycle_management ? 1 : 0
+
+  bucket = aws_s3_bucket.primary.id
+
+  rule {
+    id = var.lifecycle_transitioin_id
+
+    status = var.bucket_versioning_configuration_status
+
+    transition {
+      days          = 30
+      storage_class = var.transition_30_storage_class
+    }
+
+    transition {
+      days          = 90
+      storage_class = var.transition_90_storage_class
+    }
+
+  }
```

This dynamic rule allows you to combine the two resources into one but still allow for the toggling of `enable_storage_lifecycles`

```suggestion
    dynamic "rule" {
      for_each = var.enable_storage_lifecycles ? [1] : []
      content {
        id = "transition-storage-lifecycle"
        status = "ENABLED"
        transition {
          days          = 30
          storage_class = "STANDARD_IA"
        }
    
        transition {
          days          = 90
          storage_class = "GLACIER"
        }
      } 
    }

  }
```

---

One of the considerations they mention in the documentation is to add a `NonCurrentVersionTransition` rule if versioning is enabled. This can be done the same way as the dynamic one here and just check if versioning is enabled. https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-and-other-bucket-configs.html